docs/references/architecture/adr-117-multi-curve-support.md

```diff
@@ -0,0 +1,245 @@
+
+* **[Optional]** Use `Keyring` to load and instantiate validators when booting up a node.
+
+
+#### Create a single implementation for `PrivValidator`
+
+The current CometBFT codebase includes the following implementations of `PrivValidator`:
+
+* `FilePV`: Handles file-based private validators.
+* `SignerClient`: Manages remote signing.
+* `RetrySignerClient`: Provides retry mechanisms for remote signing.
+* `MockPV`: Used exclusively for testing purposes.
+
+We propose introducing a new implementation, `CryptoProviderPV`, which will unify and replace all the above implementations. This single implementation will act as an abstraction layer for the `PrivValidator` implementations mentioned above.
+
+**Current design:**
+
+```mermaid
+classDiagram
+    class PrivValidator {
+        <<interface>>
+    }
+
+    class FilePV {
+    }
+
+    class SignerClientPV {
+    }
+
+    class RetrySignerClientPV {
+    }
+
+    class MockPV {
+    }
+
+    PrivValidator <|.. FilePV
+    PrivValidator <|.. SignerClientPV
+    PrivValidator <|.. RetrySignerClientPV
+    PrivValidator <|.. MockPV
+```
+
+
+**Proposed design:**
+
+```mermaid
+classDiagram
+    class PrivValidator {
+        <<interface>>
+    }
+
+    class CryptoProvider {
+        <<interface>>
+    }
+
+    class CryptoProviderPV {
+    }
+
+    class FileCP {
+    }
+
+    class SocketSignerCP {
+    }
+
+    class RetrySocketSignerCP {
+    }
+
+    class MockCP {
+    }
+
+    PrivValidator <|.. CryptoProviderPV
+    CryptoProviderPV --> CryptoProvider
+    CryptoProvider <|.. FileCP
+    CryptoProvider <|.. SocketSignerCP
+    CryptoProvider <|.. RetrySocketSignerCP
+    CryptoProvider <|.. MockCP
+```
+
+For these new implementations, the current code for `File`, `SocketClient`, and `RetrySocketClient` will have to implement the `CryptoProvider` interface instead of the `PrivValidator` one.
+
+##### Code snippets for `CryptoProviderPV`
+
+As mentioned above, instead of having several implementations of `PrivValidator`, the proposal is to have only one that, by dependency injection, loads the corresponding `CryptoProvider` that offers the same functionality as the previous implementations of `PrivValidator`.
+
+Below is an example of how `CryptoProviderPV` would look like. Note that in this particular case, since the PrivateKey is managed inside the corresponding implementation, we're not passing that value to the signer. This is to avoid having to significantly change the code for `FilePV`. This is also valid for all implementations that manage their private keys in their own logic.
+
+```go
+// CryptoProviderPV is the implementation of PrivValidator using CryptoProvider's methods
+type CryptoProviderPV struct {
+    provider CryptoProvider
+    pubKey   PubKey
+}
+
+// NewCryptoProviderPV creates a new instance of CryptoProviderPV
+func NewCryptoProviderPV(provider CryptoProvider, pk PubKey) (*CryptoProviderPV, error) {
```

Comment:
Where is `pk` coming from in this use case?

Reply:
That's an error, thanks for pointing it out: keys are intended to be managed inside the crypto provider.
I see that in the code, most of the times, the pubkeys are retrieved to get the proper address. We can eventually add a `GetAddress() string` or `GetPubKey []byte` to the crypto provider interface to be able to still get that data.